### PR TITLE
HPCC-29917 Refactor compressToBuffer to support different compression methods

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jlzw.cpp [hunk 1]
```diff
@@ -2862,44 +2858,88 @@ IPropertyTree *getBlockedFileDetails(IFile *file)
...
+    }
+    bool removeCompressor(ICompressHandler *handler)
+    {
+        CompressionMethod method = handler->queryMethod();
+        if (registered.zap(* handler))
+        {
+            if ((method & ~COMPRESS_METHOD_AES) < COMPRESS_METHOD_LAST)
+            {
+                if (method & COMPRESS_METHOD_AES)
+                    AESbyMethod[method & ~COMPRESS_METHOD_AES] = handler;
```
Comment (ghalliday) [inline system/jlib/jlzw.cpp:2917]: = nullptr
and below.

File: system/jlib/jlzw.cpp [hunk 2]
```diff
@@ -758,82 +758,76 @@ size32_t RLEExpand(void *dst,const void *src,size32_t expsize)
...
+        // Alternative would be a separate enum for compressToBuffer formats, but that seems more likely to cause confusion
+        out.append((byte) (method == COMPRESS_METHOD_LZW ? COMPRESS_METHOD_LZWLEGACY : method));
+        out.append((size32_t)0);
         size32_t newSize = len * 4 / 5; // Copy if compresses less than 80% ...
-        Owned<ICompressor> compressor = createLZWCompressor();
+        Owned<ICompressor> compressor = handler->getCompressor(options);
         void *newData = out.reserve(newSize);
-        compressor->open(newData, newSize);
-        if (compressor->write(src, len)==len)
+        try
```
Comment (ghalliday) [inline system/jlib/jlzw.cpp:781]: Future: Could use compressToBlock() if supported by the compressor.  There is also a few bytes wasted, which would be more significant for small sizes, but that requires substantial refactoring.

File: testing/unittests/unittests.cpp [hunk 3]
```diff
@@ -1052,5 +1052,79 @@ class RelaxedAtomicTimingTest : public CppUnit::TestFixture
...
+        testOne(len, COMPRESS_METHOD_LZ4);
+        testOne(len, (CompressionMethod) (COMPRESS_METHOD_LZW|COMPRESS_METHOD_AES), "0123456789abcdef");
+    }
+
+    void testRun()
+    {
+        testSome(0);
+        testSome(1);
+        testSome(16);
+        testSome(32);
```
Comment (ghalliday) [inline testing/unittests/unittests.cpp:1120]: Adding testSome(48) or 64 causes a realloc error.

**Review by ghalliday — COMMENTED** [review]
One comment, and one issue to investigate

File: system/jlib/jfcmp.hpp [hunk 4]
```diff
@@ -72,8 +72,8 @@ class jlib_decl CFcmpCompressor : public CSimpleInterfaceOf<ICompressor>
 
     virtual void open(void *buf,size32_t max) override
     {
-        if (max<1024)
-            throw MakeStringException(-1,"CFcmpCompressor::open - block size (%d) not large enough", max);
+//        if (max<1024)
+//            throw MakeStringException(-1,"CFcmpCompressor::open - block size (%d) not large enough", max);
```
Comment (jakesmith) [inline system/jlib/jfcmp.hpp:76]: would be good to delete if no longer applicable (and similarly on line 106-7)

File: system/jlib/jlzw.cpp [hunk 5]
```diff
@@ -758,82 +758,76 @@ size32_t RLEExpand(void *dst,const void *src,size32_t expsize)
...
-    out.append((size32_t)0);
-
-    if (len >= 32)
+    if (method != COMPRESS_METHOD_NONE && len >= 32)
     {
+        ICompressHandler *handler = queryCompressHandler(method);
+        if (!handler)
+        {
+            VStringBuffer s("Unknown compression method %x requested in compressToBuffer", (byte) method);
+            throw makeStringException(0, s.str());
```
Comment (jakesmith) [inline system/jlib/jlzw.cpp:769]: trivial: use makeStringExceptionV?

File: system/jlib/jlzw.cpp [hunk 6]
```diff
@@ -758,82 +758,76 @@ size32_t RLEExpand(void *dst,const void *src,size32_t expsize)
...
+                size32_t compressedLen = compressor->buflen();
+                out.setWritePos(originalLength + sizeof(byte));
+                out.append(compressedLen);
+                out.setWritePos(originalLength + sizeof(byte) + sizeof(size32_t) + compressedLen);
+                return;
+            }
         }
+        catch (IException *E)
+        {
+            E->Release();
```
Comment (jakesmith) [inline system/jlib/jlzw.cpp:796]: should there a debug log of the exception here?

File: system/jlib/jlzw.cpp [hunk 7]
```diff
@@ -758,82 +758,76 @@ size32_t RLEExpand(void *dst,const void *src,size32_t expsize)
...
-            out.append(compressedLen);
-            out.setWritePos(originalLength + sizeof(bool) + sizeof(size32_t) + compressedLen);
-            return;
+            compressor->open(newData, newSize);
+            if (compressor->write(src, len)==len)
+            {
+                compressor->close();
+                size32_t compressedLen = compressor->buflen();
+                out.setWritePos(originalLength + sizeof(byte));
+                out.append(compressedLen);
```
Comment (jakesmith) [inline system/jlib/jlzw.cpp:789]: trivial: could use a DelayedMarker<size32_t> instead of line 777, and a marker.write(compressedLen) here

**Review by jakesmith — COMMENTED** [review]
@richardkchapman - looks good, some trivial comments.

**Review by jakesmith — APPROVED** [review]

File: system/jlib/jlz4.cpp [hunk 8]
```diff
@@ -73,7 +72,7 @@ class CLZ4Compressor final : public CFcmpCompressor
         if (toflush == 0)
             return;
 
-        if (toflush < 256)
+        if (false && toflush < 256)
```
Comment (ghalliday) [inline system/jlib/jlz4.cpp:75]: Should clean up and delete this code.

File: system/jlib/jlzw.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jlib/jlzw.cpp:781]: I have created a separate jira for this HPCC-30928 (and a branch gch/issue30928).  It has two effects
i) it compresses better due to the daft way the current lz4 compress works.
ii) it is faster.
Before:
```
12:24:58.258158 419929 compressToBuffer 5 size 0 did not compress
12:24:58.258387 419929 compressToBuffer 5 size 102 compressed to 66 in 1985ns
12:24:58.258803 419929 compressToBuffer 5 size 1000 compressed to 53 in 1503ns
```
After:
```
12:25:21.154246 420137 compressToBuffer 5 size 0 did not compress
12:25:21.154313 420137 compressToBuffer 5 size 34 compressed to 32 in 2434ns
12:25:21.154783 420137 compressToBuffer 5 size 1000 compressed to 35 in 1082ns
```
We could remove 4 bytes from the size, but that would be trickier (the format wouldn't be compatible).

With that change, switching to a less repetitive text  (Gen 1 from 0drvb10), LZ4 is better at compressing, but significantly slower:

```
12:37:32.829295 422255 compressToBuffer 2 size 512 compressed to 357 in 5490ns
12:37:32.837024 422255 compressToBuffer 2 size 1000 compressed to 627 in 9808ns
12:37:32.838347 422255 compressToBuffer 5 size 512 compressed to 411 in 1893ns
12:37:32.841931 422255 compressToBuffer 5 size 1000 compressed to 717 in 2885ns
```

File: system/jlib/jlzw.cpp (hunk 6 above)
Comment (ghalliday) [reply system/jlib/jlzw.cpp:796]: I don't think so - it will only happen for aes encryption of small buffers.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman one problem with sanity check, but if that is fixed it looks ok to merge.
Please can you target it to 9.4.x.  (I agree it is too painful to merge further back.)

**Review by ghalliday — CHANGES_REQUESTED** [review]
Unfortunately this has created a regression.  If you create workunits (and deploy to roxie) with the old code, and then read with the new code the system, crashes.

It is beause 
void decompressToBuffer(MemoryBuffer & out, const void * src)

does not have a leading version - it is processed in decompressResource.  I think the correct fix is to pass version (and len) from decompress resource into a shared function.

**Review by ghalliday — APPROVED** [review]
@richardkchapman please squash.

**Comment by richardkchapman** [comment]
@ghalliday Squashed

**Pull request merged**